### DISCUSSION TITLE
sqlalchemy unfathomably slow on sqlite but not on postgres

### DISCUSSION THREAD

**Discussion by epigramx** [sqlalchemy/sqlalchemy#7997] [Usage Questions]
I'm not very familiar with the project that uses it to be honest but it appears to be regular sqlalchemy-ORM usage and the main dev reports "_we use the implicitly provided query_property() from the session - which is based on the sqlalchemy documentaiton_".

Issue is for the same use-case: sqlite is more than ~50X slower than dumping it to a postgres db and using that instead; I guess it might be normal but it sounds extreme that one takes 3sec and the other 3min; I wonder if it rings any bells to you.

An issue at that project that mentions it and some on the surface discussion https://github.com/freqtrade/freqtrade/issues/6791

**Comment by zzzeek** [comment]
no bells here, there's a lot of things that can be going on.   Work through the steps at https://docs.sqlalchemy.org/en/14/faq/performance.html#how-can-i-profile-a-sqlalchemy-powered-application which should quickly shed some light.  sounds like number of queries being emitted.  if you have a very large bulk insert, that might be much faster on Postgresql as we have some optimization paths in place for that.

**Reply by epigramx** [reply]
In this case it appears the queries shown with echo=True aren't that many to explain it. The CPU load is very high for the python process during the stall though which is as if something is busy-looping. I pause in vs code and see nothing relevant on the client software so it's like something at the sqlite level is busy-looping (especially since the CPU load goes away in less than 5 seconds on postgres).

**Reply by zzzeek** [reply]
do you have a query that is not well optimized, or any custom Python sqlite functions in use?   when you use cprofile, it will show it taking a long time on ``cursor.execute()`` which is where it hits pysqlite.

**Reply by xmatthias** [reply]
The problematic query is using limit/offset (with increasing offsets, as it's a paginated endpoint).

in my understanding, this is something that sqlite does not properly support (with an offset of 6000, it'll have to fetch all 6000 prior rows, to get to rows 6000-6500 (limit is 500)).
The query itself is a plain select with a filter on an indexed column - and a relationship join (left or inner doesn't make a difference to performance) on to additional table, which contains several rows per primary row.

One potential problem is the query that's coming out.

simplified, it's 

``` sql
select anon_1.a, anon_1.b, anon_1.c --...
from (trades .a, trades .b, trades .c -- X more columns)
from trades 
where trades.b is 0 order by trades.a
limit 500 offset 6500) as anon_1 left outer join orders as orders_1 ON anon_1.trades_id = orders_1.ft_trade_id ORDER BY anon_1.trades_id, orders_1.id
```

running the above query results however in ~4000 objects in total - the result will be 500 trade objects - each with several orders attached. 

The performance difference is also noticable in a regular query program (sqlitebrowser) - between offset 0  and offset 6500 there's quite some difference in query runtime between offset 0 and offset 6000

Run Time: real 0.009 user 0.008341 sys 0.000000 -- Offset 0
Run Time: real 0.071 user 0.047692 sys 0.015896 --offset 6500

Using a inner join (instead of the left outer join) doesn't change performance (nor the result) - as all columns involved as filter or join criteria are indexed or the primary key.

**Reply by epigramx** [reply]
Is this relevant? https://stackoverflow.com/a/30124574/277716 
![image](https://user-images.githubusercontent.com/2960257/167566728-893c81e3-c8e7-469c-b463-d9cbef98c846.png)
It appears there's no (enumeration)index in this particular dbase so it sounds like one could be used for the purpose but I don't know if something in the relational nature of the tables would prevent that use case.

**Reply by xmatthias** [reply]
it's a possibility (one i experimented with) - but it doesn't improve performance in the above scenario with my 7k database significantly.
we'll still have to use `.limit` to ensure we get the correct number of rows (id's can be missing in the middle). While that's not a problem - it didn't improve performance, either).

while there is no specific index - the primary key is apparently [automatically indexed](https://stackoverflow.com/questions/3379292/is-an-index-needed-for-a-primary-key-in-sqlite) (it does however not change the performance of the requests).

**Reply by epigramx** [reply]
There is an index but it can not be used to replace offsetting because if an entry is deleted: that index remains at its original value, so a new column would have to be used and repopulated every time a new record changes which holds the exact number of rows on the table.
I don't know why you say it didn't improve performance since you said sqlite took 2.3 seconds for 500 entries in some cases and it makes sense that it would be faster because a `WHERE` goes exactly where it needs to be instead of `OFFSET` getting old unrelated records.

**Reply by xmatthias** [reply]
> I don't know why you say it didn't improve performance since you said sqlite took 2.3 seconds for 500 entries in some cases and it makes sense that it would be faster because a WHERE goes exactly where it needs to be instead of OFFSET getting old unrelated records.

Because i tested with it - and experienced it as "just as slow". it makes sense to me too that it would be faster - but it wasn't.

**Reply by zzzeek** [reply]
> The problematic query is using limit/offset (with increasing offsets, as it's a paginated endpoint).

oh . well yes, those don't scale at all.  your PG version will also grind to a halt someday, just FYI I've been there on this one.

> 
> in my understanding, this is something that sqlite does not properly support (with an offset of 6000, it'll have to fetch all 6000 prior rows, to get to rows 6000-6500 (limit is 500)).

IIUC Postgresql does as well, unless they've implemented something magical recently.   OFFSET is basically useless for arbitrary N.   Here's a post about this from a guy that writes about this a lot: https://use-the-index-luke.com/blog/2013-07/pagination-done-the-postgresql-way 

that post is from 2013.  On our end we also propose using window functions for pagination (SQLite supports these), see https://github.com/sqlalchemy/sqlalchemy/wiki/RangeQuery-and-WindowedRangeQuery .


basically, dont use OFFSET, it won't perform well anywhere.

**Reply by xmatthias** [reply]
Using the "id" filtering method for now - it's not really better - and "later" results (with higher ID) still have much lower performance.

The query boils down to the following - with id being a primary key (i've now defined it with index, too - but that dosn't seem to change much).

```
Trade.query.filter([Trade.id > start_id]).order_by(Trade.id).limit(limit)
```

in my expectation, timing should be "mostly" linear - but it really isn't.

---

Removing the `lazy="joined"` (so the relationship is not resolved at all) from the relationship significantly improves performance (into <100ms ranges) - from slowly degrating per cycle up to 2 s with joined loading).

Both `lazy='subquery'` and `lazy='selectin'` yield better results than joined, too - but in my understanding of the [docs](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html) - both will issue a 2nd statement - whereas a join statement (on indexed keys!) is usually the best option when looking at the database alone (detached from sqlalchemy).

**Reply by zzzeek** [reply]
Im sure sqlite has a mailing list where you can get a lot of help optimizing a query.

**Reply by xmatthias** [reply]
it's sqlalchemy that generates the query through the ORM (see the above ORM statement).

I suspect the following will be the problematic point: https://docs.sqlalchemy.org/en/14/tutorial/orm_related_objects.html#joined-load

> [joinedload()](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#sqlalchemy.orm.joinedload) also works for collections, meaning one-to-many relationships, however it has the effect of multiplying out primary rows per related item in a recursive way that grows the amount of data sent for a result set by orders of magnitude for nested collections and/or larger collections, so its use vs. another option such as [selectinload()](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#sqlalchemy.orm.selectinload) should be evaluated on a per-case basis.

Although the documentation does not clarify it for me if `lazy='joined'` is the same as the `joinedload()` mentioned above.

**Reply by CaselIT** [reply]
> Although the documentation does not clarify it for me if `lazy='joined'` is the same as the `joinedload()` mentioned above.

yes, it's the same.

> Both lazy='subquery' and lazy='selectin' yield better results than joined, too - but in my understanding of the [docs](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html) - both will issue a 2nd statement - whereas a join statement (on indexed keys!) is usually the best option when looking at the database alone (detached from sqlalchemy).

the second query should not really be an issue with sqlite, since it usually has a very low latency compared to server side db like postgres since it does not need to go over the network

**Reply by zzzeek** [reply]
> it's sqlalchemy that generates the query through the ORM (see the above ORM statement).

you have complete control over the structure of a statement.    A statement like `Trade.query.filter([Trade.id > start_id]).order_by(Trade.id).limit(limit)` is unremarkable, so if there is a performance issue with that, you'd need to view the SQL being sent to the database and review what's wrong with it from a SQL perspective.   

> 
> I suspect the following will be the problematic point: https://docs.sqlalchemy.org/en/14/tutorial/orm_related_objects.html#joined-load
> 
> > [joinedload()](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#sqlalchemy.orm.joinedload) also works for collections, meaning one-to-many relationships, however it has the effect of multiplying out primary rows per related item in a recursive way that grows the amount of data sent for a result set by orders of magnitude for nested collections and/or larger collections, so its use vs. another option such as [selectinload()](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#sqlalchemy.orm.selectinload) should be evaluated on a per-case basis.

`joinedload()` uses a JOIN or OUTER JOIN (you can select which you'd prefer) and if you set logging level to DEBUG, it will illustrate all the rows being returned.  you can switch this to `selectinload()` which for collections is generally much more efficient than `joinedload()`.  Background on this is at https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html 

> 
> Although the documentation does not clarify it for me if `lazy='joined'` is the same as the `joinedload()` mentioned above.

this is explicitly documented at https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html, fifth paragraph, second bullet

**Reply by epigramx** [reply]
As I said I'm not very familiar with that project's code but: I just deleted a `.limit(limit).offset(offset)` that was at a relevant `order_by(order_by).limit(limit).offset(offset)` it was using and now it (correctly)processes results almost instantly(!). I noticed that in the wiki page posted above (https://github.com/sqlalchemy/sqlalchemy/wiki/RangeQuery-and-WindowedRangeQuery): it mentions that the need to have pages could be related to psycopg2's server-side limitations but that removal didn't have issues with PG either (and it also made it faster). I guess I don't know when pages are even needed but in this case it **appears** they are not.

**Reply by CaselIT** [reply]
It seems very strange that removing limit/offset would make the query faster when loading all the results, since the db has to do *more* work, not less.

Could you share what sql is being executed with and without the limit/offset?

**Reply by epigramx** [reply]
@CaselIT not sure why you think it would do more work. Wasn't it clarified above that `OFFSET` is severely inefficient, because each iteration it(the db) also scans though all the older and unrelated(unneeded) results?

**Reply by CaselIT** [reply]
> @CaselIT not sure why you think it would do more work.

well in case of limit - offset the db has to
- discard rows until offset ones have been discarded
- return limit rows

this is less work than returning all rows from the table

>  Wasn't it clarified above that OFFSET is severely inefficient, because each iteration it(the db) also scans though all the older and unrelated(unneeded) results?

sure but if you replace a query with limit offset with the same query without, as you mentioned above, the db will have to do more work

**Reply by epigramx** [reply]
@CaselIT the user doesn't seem to be looping there, so without `limit.offset` the db is probably getting all records at once and be done with it. This is the function where they're using it by the way: https://github.com/freqtrade/freqtrade/blob/891900c186a75d537a79c3adbf82ed17afaa379e/freqtrade/rpc/rpc.py#L413 (edit: and that appears to be used in return by a website frontend here https://github.com/freqtrade/frequi/blob/ccfa81b05152c1154aeb56494a32738bebd061f4/src/stores/ftbot.ts#L253 which encounters the original slow-down).

**Reply by CaselIT** [reply]
>  the db is probably getting all records at once and be done with it.

should not really change with a limit offset, since from the db point of view is exactly the same thing, but stops at limit+offset. 

In any case whatever works...
Note that if you have 10M rows probably the caller is not happy to get them all back at once

**Reply by epigramx** [reply]
It doesn't stop in that implementation though (second URL at previous reply of mine) so it either gets them all at once (with my hack) or it gets them all with limit+offset and it's much slower.
They have a comment in there: "_would fire all requests at once, which can cause problems for big sqlite databases_", and I've no clue how it's justified since why would **sqlite** care?

**Reply by CaselIT** [reply]
I'm no sqlite expert, so don't know. Just observing that logically that behaviour is strange.
Then again db are strange sometimes, so whatever makes it happy I guess

**Reply by epigramx** [reply]
I seriously doubt that sqlite itself cares if a list of results is long so it's mainly a question of if the project that uses it can handle a long list of results. That project recently fixed a longstanding concurrency/multithreading bug by using `PRAGMA journal_mode=wal`, and I suspect that whole thing ("_can cause problems with sqlite if we fire a lot requests at once_" which lead them to use pagination and `OFFSET`) was an earlier attempt to workaround that old multithreading bug and it was left in accidentally.

**Reply by xmatthias** [reply]
@epigramx you're making things look in a way they are not - which is missleading people trying to help here.

obviously, doing ONE query for 10k rows is faster than 20 roundtrips with limit/offset in batches of 500 - at least from a database perspective.
That's not something anyone will disagree with - as it's 1 query vs. 20 query. You can optimize the query as long as you want, 1 query will always be faster than 20 queries.

but doing one query for all rows will have other side-effects (huge response sizes) - which is something that's problematic in other circumstances - and can cause problems with json serialize (especially on systems with low memory).

**Reply by epigramx** [reply]
> doing one query for all rows will have other side-effects (huge response sizes) - which is something that's problematic in other circumstances - and can cause problems with json serialize (especially on systems with low memory)

Your current code grinds a modern PC to a halt, while not using pagination at all lets it respond almost instantly without issues. If you are convinced there will be issues in some systems and you are convinced pagination with tracking ids is not better than offset then you should identify what the system limitations are because it doesn't make sense to grind a modern system to a halt for the promise it might run out of resources or for the promise it might have other side-effects that don't otherwise appear.
